blockchain/src/codec.rs

```diff
@@ -0,0 +1,376 @@
+        self.read_u8_vec().map(ShortIDVec)
+    }
+
+    fn read_blockid(&mut self) -> Result<BlockID, ReadError> {
+        self.read_u8x32().map(BlockID)
+    }
+
+    fn read_hash(&mut self) -> Result<Hash, ReadError> {
+        self.read_u8x32().map(Hash)
+    }
+
+    fn read_block_header(&mut self) -> Result<BlockHeader, ReadError> {
+        Ok(BlockHeader {
+            version: self.read_u64()?,
+            height: self.read_u64()?,
+            prev: self.read_blockid()?,
+            timestamp_ms: self.read_u64()?,
+            txroot: self.read_hash()?,
+            utxoroot: self.read_hash()?,
+            ext: self.read_u8_vec()?,
+        })
+    }
+
+    fn read_inventory(&mut self) -> Result<Inventory, ReadError> {
+        Ok(Inventory {
+            version: self.read_u64()?,
+            tip: self.read_block_header()?,
+            tip_signature: self.read_signature()?,
+            shortid_nonce: self.read_u64()?,
+            shortid_list: self.read_shortid_vec()?,
+        })
+    }
+
+    fn read_tx(&mut self) -> Result<Tx, ReadError> {
+        let tx_size = self.read_u32()? as usize;
+        let vec = self.read_vec(tx_size)?;
+        Tx::from_bytes(&vec).map_err(|_| ReadError::InvalidFormat)
+    }
+
+    fn read_txs(&mut self) -> Result<Vec<BlockTx>, ReadError> {
+        let len = self.read_u32()? as usize;
+        const BLOCK_TX_MIN_LENGTH: usize = 8 + 8 + 8 + 8 + 32 + 32 + 1 + 11 * 32 + 8;
+        self.read_vec_with(len, BLOCK_TX_MIN_LENGTH, |src| {
+            let tx = src.read_tx()?;
+            let len = src.read_u32()? as usize;
+            let proofs = src.read_vec_with(len, 1, |src| match src.read_u8()? {
+                0 => Ok(Proof::Transient),
+                1 => merkle::Path::decode(src)
+                    .map(Proof::Committed)
+                    .map_err(|_| ReadError::InvalidFormat),
+                _ => Err(ReadError::InvalidFormat),
+            })?;
+            Ok(BlockTx { tx, proofs })
+        })
+    }
+}
+
+trait WriterExt: Writer + Sized {
+    fn write_u8_vec(&mut self, label: &'static [u8], vec: &[u8]) -> Result<(), WriteError> {
+        self.write_u32(b"vec length", vec.len() as u32)?;
+        self.write(label, vec)?;
+        Ok(())
+    }
+
+    fn write_signature(&mut self, sig: &Signature) -> Result<(), WriteError> {
+        self.write(b"scalar", sig.s.as_bytes())?;
+        self.write(b"key", sig.R.as_bytes())?;
+        Ok(())
+    }
+
+    fn write_shortid_vec(
+        &mut self,
+        label: &'static [u8],
+        vec: &ShortIDVec,
+    ) -> Result<(), WriteError> {
+        self.write_u8_vec(label, vec.0.as_ref())
+    }
+
+    fn write_blockid(
+        &mut self,
+        label: &'static [u8],
+        block_id: &BlockID,
+    ) -> Result<(), WriteError> {
+        self.write(label, block_id.as_ref())
+    }
+
+    fn write_hash(&mut self, label: &'static [u8], hash: &Hash) -> Result<(), WriteError> {
+        self.write(label, hash.0.as_ref())
```

Comment:
there's `impl Deref for Hash` with `Target = [u8]`, so we don't even need this helper. Just write `write(b"foo", &hash)`.

Reply:
Oh, let's keep it for symmetry with `read_hash`, which is nice because it does the type wrapping.

Reply:
Added impl Decodable/Encodable